prosthesis/components/SimulatorScreen.js

```
@@ -27,16 +38,113 @@ SimulatorScreen.prototype = {
     flags: Ci.nsIClassInfo.DOM_OBJECT
   }),
 
-  init: function (aWindow) {
-    dump("SIMULATOR SCREEN INIT CALLED\n");
+  _getOrigin: function(aURL) {
+    let uri = Services.io.newURI(aURL, null, null);
+    return uri.prePath;
+  },
+
+  _createEvent: function(name) {
+    return new this._aWindow.Event(name);
+  },
+
+  _fireMozOrientationChangeEvent: function() {
+    try {
+      let e = this._createEvent("mozorientationchange");
+      if (this._onMozOrientationChange) {
+        this._onMozOrientationChange(e);
+      }
+      this.dispatchEvent(e);
+    } catch(e) {
+      debug(["EXCEPTION:", e, e.fileName, e.lineNumber].join(' '));
+    }
+  },
+
+  dispatchEvent: function(evt) {
+    if (!this._eventListenersByType) {
+      return;
+    }
+
+    let type = evt.type;
+    var listeners = this._eventListenersByType[type];
+    if (listeners) {
+      for (let i = 0, len = listeners.length; i < len; i++) {
+        let listener = listeners[i].listener;
+
+        try {
+          if (typeof listener == "function") {
+            listener.call(this, evt);
+          } else if (listener && listener.handleEvent &&
+                     typeof listener.handleEvent == "function") {
+            listener.handleEvent(evt);
+          }
```

_listener_ must be a function or an object with a _handleEvent_ method, so this shouldn't check that _handleEvent_ is callable before calling it and ignore it if isn't. It would be better for the code to simply throw, informing the developer about the problem, than to fail silently, i.e.:

```
      if (typeof listener == "function") {
        listener.call(this, evt);
      } else {
        listener.handleEvent(evt);
      }
```